### PR TITLE
Fix IllegalStateException in PersistentReplicator

### PR DESCRIPTION
**Pull request by lhotari**
Fixes: #10097 

### Motivation

See #10097 for the issue. It seems that the code broke when the switch was made to LightProto in #9046.

### Modifications

It is necessary to use `msg.getMessageBuilder().hasReplicatedFrom()` and use logic that only calls `msg.getMessageBuilder().getReplicatedFrom()` if `hasReplicatedFrom()` returns true.

### CODE REVIEW

**Review by eolivelli — APPROVED** [review]
good catch.
the change looks good to me

Do you mind adding some minimal test case that reproduces the problem ?

it is scary that we do not have any broken test.

My reasoning is that we can commit this fix as it is but, as we are not in a hurry (no need for an hotfix, as 2.8.0 is still unreleased), we can spend a little more time and add a test that covers this change.

**Comment by lhotari** [comment]
> Do you mind adding some minimal test case that reproduces the problem ?
> 
> it is scary that we do not have any broken test.
> 
> My reasoning is that we can commit this fix as it is but, as we are not in a hurry (no need for an hotfix, as 2.8.0 is still unreleased), we can spend a little more time and add a test that covers this change.

Makes sense. It seems that integration tests are missing for replicated subscriptions (PIP-33) so it requires wider understanding of the feature to be able to add the required test coverage. 

@merlimat do you have recommendations of how replicated subscriptions could be tested as part of the automated tests in CI?

**Comment by lhotari** [comment]
Fixing the IllegalStateException doesn't fix subscription replication. A new exception appears. I reported that in https://github.com/apache/pulsar/issues/10097#issuecomment-810997397 .

**Comment by lhotari** [comment]
I added a test that reproduces #10054 and it also reproduced the IllegalStateException . However it didn't reproduce [the IllegalReferenceCountException that I noticed when testing manually](https://github.com/apache/pulsar/issues/10097#issuecomment-810997397).

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    public void cleanup() throws Exception {
+        super.cleanup();
+    }
+
+    /**
+     * Tests replicated subscriptions across two regions
+     */
+    @Test
+    public void testReplicatedSubscriptionAcrossTwoRegions() throws Exception {
+        String namespace = "pulsar/replicatedsubscription-" + System.currentTimeMillis();
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:66]: there is some utility method to create a unique name, can you please check ?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        try (Consumer<byte[]> consumer1 = client1.newConsumer()
+                .topic(topicName)
+                .subscriptionName(subscriptionName)
+                .replicateSubscriptionState(replicateSubscriptionState)
+                .subscribe()) {
+            readMessages(consumer1, receivedMessages, 3, allowDuplicates);
+        }
+
+        // wait for subscription to be replicated
+        Thread.sleep(2 * config1.getReplicatedSubscriptionsSnapshotFrequencyMillis());
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:123]: is there any way to use awaiatility ?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java (hunk 2 above)
Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:123]: The replication should happen with the frequency configured by "replicatedSubscriptionsSnapshotFrequencyMillis" (1000ms default, also set in the test). I'm expecting that 2 * replicatedSubscriptionsSnapshotFrequencyMillis is a sufficient delay.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java (hunk 1 above)
Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:66]: that's true. I copied this method from another ReplicatorTest.

Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java]: I pushed a fix using `BrokerTestUtil.newUniqueName`

**Review by eolivelli — APPROVED** [review]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentReplicator.java [hunk 3]
```diff
@@ -760,7 +760,7 @@ private void checkReplicatedSubscriptionMarker(Position position, MessageImpl<?>
 
         int markerType = msg.getMessageBuilder().getMarkerType();
 
-        if (!remoteCluster.equals(msg.getMessageBuilder().getReplicatedFrom())) {
+        if (!remoteCluster.equals(msg.getReplicatedFrom())) {
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentReplicator.java:763]: I am not sure this is the right fix. 

If you want to address the issue `java.lang.IllegalStateException: Field 'replicated_from' is not set`, you should check `msg.getMessageBuilder().hasReplicatedFrom()`.

Your change brings inconsistent usage across different places. Line 756 and 761 are using `MessageBuilder` and Line 763 is using `Message`.

**Review by sijie — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentReplicator.java (hunk 3 above)
Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentReplicator.java]: Changed as suggested.

**Comment by eolivelli** [comment]
@sijie can you please take another look ?

**Review by merlimat — APPROVED** [review]
👍

**Review by codelipenghui — APPROVED** [review]

**Review by sijie — APPROVED** [review]

**Comment by eolivelli** [comment]
LightProto is not present in branch-7.2, removing release/2.7.2 label

**Pull request merged**